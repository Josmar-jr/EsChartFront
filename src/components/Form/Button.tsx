import { ButtonHTMLAttributes, ReactNode } from 'react';

import { CircleNotch } from 'phosphor-react';

export type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> & {
  isLoading?: boolean;
  children: ReactNode;
};

export function Button({ isLoading = false, children, ...rest }: ButtonProps) {
  return (
    <button
      type="submit"
      className="disabled:bg-primaryDark dark:bg-secondary disabled:cursor-not-allowed group relative w-full flex justify-center py-2 px-4 border border-primary dark:border-secondary text-sm font-medium rounded-md text-white bg-primary hover:brightness-90 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary focus:dark:ring-secondary"
      disabled={isLoading}
      {...rest}
    >
      {isLoading ? (
        <CircleNotch
          className="animate-spin"
          size={18}
          weight="bold"
          color="white"
        />
      ) : (
        children
      )}
    </button>
  );
}
