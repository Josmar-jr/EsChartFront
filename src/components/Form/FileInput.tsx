import { Plus } from 'phosphor-react';
import {
  useState,
  SetStateAction,
  Dispatch,
  ForwardRefRenderFunction,
  forwardRef,
  useCallback,
  useEffect
} from 'react';
import {
  FieldError,
  FieldValues,
  UseFormSetError,
  UseFormTrigger
} from 'react-hook-form';

export interface FileInputProps {
  name: string;
  error?: FieldError;
  setImageUrl: Dispatch<SetStateAction<string>>;
  localImageUrl: string;
  setLocalImageUrl: Dispatch<SetStateAction<string>>;
  setError: UseFormSetError<FieldValues | any>;
  onChange: (
    event: React.ChangeEvent<HTMLInputElement>
  ) => Promise<boolean | void>;
  trigger: UseFormTrigger<FieldValues | any>;
}

const FileInputBase: ForwardRefRenderFunction<
  HTMLInputElement,
  FileInputProps
> = (
  {
    name,
    error = null,
    setImageUrl,
    localImageUrl,
    setLocalImageUrl,
    setError,
    onChange,
    trigger,
    ...rest
  },
  ref
) => {
  const [isVisibleTextHover, setIsVisibleTextHover] = useState(false);
  const [progress, setProgress] = useState(0);
  const [isSending, setIsSending] = useState(false);

  const handleImageUpload = useCallback(
    async (event: React.ChangeEvent<HTMLInputElement>): Promise<void> => {
      if (!event.target.files?.length) {
        return;
      }

      setImageUrl('');
      setLocalImageUrl('');
      setError('image', null);
      setIsSending(true);

      await onChange(event);
      trigger('image');

      const formData = new FormData();

      formData.append(event.target.name, event.target.files[0]);
      setLocalImageUrl(URL.createObjectURL(event.target.files[0]));
      console.log('ttt', localImageUrl);
      setIsSending(false);
    },
    [onChange, setError, localImageUrl, setImageUrl, setLocalImageUrl, trigger]
  );

  return (
    <label htmlFor={name} className="relative h-44 w-44 items-center mx-auto block cursor-pointer">
      {localImageUrl && !isSending ? (
        <div className="rounded-full w-full h-full relative cursor-pointer border-4 border-primary">
          <img
            src={localImageUrl}
            className="rounded-full w-full h-full object-cover"
            alt="Uploaded photo"
          />
          <div
            onMouseLeave={() => setIsVisibleTextHover(false)}
            onMouseEnter={() => setIsVisibleTextHover(true)}
            className={`absolute ${
              isVisibleTextHover ? 'opacity-100' : 'opacity-0'
            } hover-text z-10 bg-[rgba(0,0,0,0.2)] transition-opacity flex items-center justify-center`}
          >
            <span className={`${isVisibleTextHover ? 'translate-y-12' : 'translate-y-60'} transition-transform block text-xs text-slate-200 text-center pt-2 bg-[rgba(0,0,0,0.8)]`}>
              Click para alterar o avatar
            </span>
          </div>
        </div>
      ) : (
        <div className="w-full h-full flex-col items-center justify-center border-2 border-primary rounded-full bg-green-200 color-gray-200">
          <div className="w-full h-full">
            <div className="flex h-full items-center justify-center flex-col">
              <Plus size={32} weight="bold" />
              <span className="text-center pt-2">Adicione sua imagem</span>
            </div>
          </div>
        </div>
      )}

      <input
        data-testid={name}
        disabled={isSending}
        id={name}
        name={name}
        onChange={handleImageUpload}
        ref={ref}
        type="file"
        className="hidden"
        {...rest}
      />
    </label>
  );
};

export const FileInput = forwardRef(FileInputBase);
