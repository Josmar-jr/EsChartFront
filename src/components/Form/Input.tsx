import * as Tooltip from '@radix-ui/react-tooltip';
import { WarningCircle } from 'phosphor-react';
import { forwardRef, ForwardRefRenderFunction } from 'react';
import { FieldError } from 'react-hook-form';

interface InputProps extends React.InputHTMLAttributes<HTMLInputElement> {
  name: string;
  error?: FieldError;
  customClass?: string;
}

const InputBase: ForwardRefRenderFunction<HTMLInputElement, InputProps> = (
  { name, customClass = '', error = null, ...rest },
  ref
) => {
  return (
    <label
      htmlFor={name}
      className="flex self-center items-center relative flex-1 bg-neutral"
    >
      <input
        id={name}
        name={name}
        className={`${customClass} appearance-none rounded-none w-full px-3 py-2 bg-neutral ${
          !!error
            ? 'border-red-600 focus:border-red-700'
            : 'border-gray-300 focus:border-primary'
        } border  placeholder-gray-500 text-gray-900 focus:z-10 focus:outline-none   sm:text-sm`}
        {...rest}
        ref={ref}
      />
      {!!error && (
        <div className="flex items-center absolute right-0 z-10">
          <Tooltip.Provider>
            <Tooltip.Root>
              <Tooltip.Trigger className="hover:text-red-600 text-red-500">
                <WarningCircle
                  size={18}
                  weight="bold"
                  className="m-2 bg-neutral"
                />
              </Tooltip.Trigger>
              <Tooltip.Content
                side="right"
                className="bg-red-100 px-2 py-1 text-xs w-26"
              >
                <Tooltip.Arrow className="fill-red-100" />
                {error.message}
              </Tooltip.Content>
            </Tooltip.Root>
          </Tooltip.Provider>
        </div>
      )}
    </label>
  );
};

export const Input = forwardRef(InputBase);
