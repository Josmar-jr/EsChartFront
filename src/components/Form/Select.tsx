import { SelectHTMLAttributes } from 'react';

type SelectProps = SelectHTMLAttributes<HTMLSelectElement> & {
  customClass?: string;
  options?: string[];
  defaultValue?: string;
};

export function Select({
  customClass,
  options,
  defaultValue,
  ...selectProps
}: SelectProps) {
  return (
    <select
      id="small"
      className="block px-3 py-2 w-full text-sm text-gray-900 
      bg-neutral-200 focus:outline-none rounded-lg border border-gray-300 sm:text-md
      focus:ring-primary focus:border-primary
      dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 
      dark:focus:border-blue-500 shadow-sm"
      {...selectProps}
    >
      {options.map(option => {
        if (option === defaultValue) {
          return (
            <option value={option} key={option} defaultValue={option}>
              {option}
            </option>
          );
        }
        return (
          <option value={option} key={option}>
            {option}
          </option>
        );
      })}
    </select>
  );
}
