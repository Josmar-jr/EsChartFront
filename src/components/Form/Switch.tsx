interface SwitchProps extends React.InputHTMLAttributes<HTMLInputElement> {
  isChecked: boolean;
  onChangeChecked: () => void;
}

const afterStyle = `after:content-[''] after:absolute after:top-[2px] after:left-[4px] after:bg-slate-800 after:border-slate-600 after:border after:rounded-full after:h-3 after:w-3 after:transition-all after:translate-x-[-10%] peer-checked:after:translate-x-[110%] peer-checked:after:bg-secondary peer-checked:after:border-secondary`;

const beforeStyle = `bg-gray-200 before:content-[''] before:h-3 before:w-[6.5px] before:rotate-45 before:absolute before:border-r-[3px] before:border-b-[3px] before:border-secondary before:top-[0px] before:left-[6px]`;

export function Switch({ isChecked, onChangeChecked, ...rest }: SwitchProps) {
  return (
    <span className="flex gap-4 items-center justify-center">
      <label htmlFor="default-toggle">
        <span className="cursor-pointer">Tema escuro</span>
      </label>
      <div
        className={`inline-flex relative items-center cursor-pointer border-2 rounded-full ${
          isChecked ? 'border-secondary' : 'border-slate-800'
        }`}
      >
        <input
          type="checkbox"
          defaultChecked={isChecked}
          id="default-toggle"
          className="sr-only peer"
          {...rest}
        />
        <div
          className={`${
            isChecked && beforeStyle
          } ${afterStyle} w-[32px] h-4 outline-none rounded-full peer dark:border-slate-800 peer-checked:bg-slate-900`}
        ></div>
      </div>
    </span>
  );
}
