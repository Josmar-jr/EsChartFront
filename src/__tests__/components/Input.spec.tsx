import { render, screen, cleanup } from '@testing-library/react';
import { Input } from '../../components/Form/Input';
import renderer from 'react-test-renderer';

afterEach(() => {
  cleanup();
});

describe('Input Component', () => {
  it('should render correctly', () => {
    render(<Input name="email" />);

    const input = screen.getByLabelText('email');

    expect(input).toBeInTheDocument();
  });

  it('should show the toast and change input border color when to error', () => {
    const mockError = {
      message: 'email must be a valid email',
      type: 'email'
    };

    render(<Input name="email" error={mockError} />);

    const tooltipAlert = screen.getByRole('alert');
    const input = screen.getByLabelText('email');

    expect(tooltipAlert).toBeInTheDocument();
    expect(input).toHaveClass('border-red-600 focus:border-red-700');
  });

  it('should match snapshot input without prop error', () => {
    const tree = renderer.create(<Input name="email" />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot input with prop error', () => {
    const mockError = {
      message: 'email must be a valid email',
      type: 'email'
    };
    const tree = renderer
      .create(<Input name="email" error={mockError} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
