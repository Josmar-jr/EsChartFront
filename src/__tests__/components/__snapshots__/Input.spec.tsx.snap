// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component should match snapshot input with prop error 1`] = `
<label
  className="flex self-center items-center relative flex-1 bg-neutral"
  data-testid="label-wrapper"
  htmlFor="email"
>
  <input
    aria-label="email"
    className=" appearance-none rounded-none w-full px-3 py-2 bg-neutral border-red-600 focus:border-red-700 border placeholder-gray-500 text-gray-900 focus:z-10 focus:outline-none sm:text-sm"
    id="email"
    name="email"
  />
  <div
    className="flex items-center absolute right-0 z-10"
    role="alert"
  >
    <button
      className="hover:text-red-600 text-red-500"
      data-state="closed"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <svg
        className="m-2 bg-neutral"
        fill="currentColor"
        height={18}
        viewBox="0 0 256 256"
        width={18}
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <circle
          cx="128"
          cy="128"
          fill="none"
          r="96"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="24"
        />
        <line
          fill="none"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="24"
          x1="128"
          x2="128"
          y1="80"
          y2="132"
        />
        <circle
          cx="128"
          cy="172"
          r="16"
        />
      </svg>
    </button>
  </div>
</label>
`;

exports[`Input Component should match snapshot input without prop error 1`] = `
<label
  className="flex self-center items-center relative flex-1 bg-neutral"
  data-testid="label-wrapper"
  htmlFor="email"
>
  <input
    aria-label="email"
    className=" appearance-none rounded-none w-full px-3 py-2 bg-neutral border-gray-300 focus:border-primary border placeholder-gray-500 text-gray-900 focus:z-10 focus:outline-none sm:text-sm"
    id="email"
    name="email"
  />
</label>
`;
